app/reconciliation.go

```diff
@@ -0,0 +1,148 @@
+package app
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+)
+
+const reconStoreKey = "reconciliation"
+
+// unbalancedBlockHeightKey for saving unbalanced block height for reconciliation
+var unbalancedBlockHeightKey = []byte("0x01")
+
+var (
+	SupplyKey          = banktypes.SupplyKey
+	DenomAddressPrefix = banktypes.DenomAddressPrefix
+	BalancesPrefix     = banktypes.BalancesPrefix
+)
+
+// reconBalance will do reconciliation for accounts balances.
+func (app *App) reconBalance(ctx sdk.Context, bankIavl *iavl.Store) {
+	if ctx.BlockHeight() <= 2 {
+		return
+	}
+
+	height, exists := app.getUnbalancedBlockHeight(ctx)
+	if exists {
+		panic(fmt.Sprintf("unbalanced state at block height %d, please use hardfork to bypass it", height))
+	}
+
+	balanced := app.getBankChanges(ctx, bankIavl)
+	bankIavl.ResetDiff()
+
+	if !balanced {
+		app.saveUnbalancedBlockHeight(ctx)
+	}
+}
+
+func (app *App) getBankChanges(ctx sdk.Context, bankIavl *iavl.Store) bool {
```

Comment:
the returned bool value is not related to the function name `getBankChanges`, can we have a better name for this function

Reply:
renamed the function